### PR TITLE
Update the contribution guide to include a applications section

### PR DESCRIPTION
**Pull request by mattdangerw**
There are a number of things to keep in mind when contributint
Keras applications. This provides a checklist of things we will
look for from contributors.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -211,3 +211,39 @@ command above.
...
+## Contributing to Keras applications
+
+Contributions to the
+[pre-trained application library](https://keras.io/api/applications/)
+are welcome. Code for Keras applications is located in Keras repository in
+[keras/applications](https://github.com/keras-team/keras/blob/master/keras/applications).
+When contributing to Keras applications, please keep following checklist in
+mind.
+
+ - Keras applications must implement an estabilished and widely used model.
```
Comment (fchollet) [inline CONTRIBUTING.md:224]: Typo: established

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -211,3 +211,39 @@ command above.
...
+   Applications should include a link to a paper describing the architecture of
+   the model with at least 20 citations.
+ - Applications should be provided with pre-trained weights.
+   - When submitting a pull request for a Keras application, these weights can
+     be provided at any publically available URL (e.g. a personal Cloud Storage
+     bucket). The weights will be uploaded to a Keras storage bucket while
+     merging the pull request.
+   - Weights should be downloaded with the
+     [get_file](https://keras.io/api/utils/python_utils/#getfile-function)
```
Comment (fchollet) [inline CONTRIBUTING.md:233]: Add parens: `get_file()`

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -211,3 +211,39 @@ command above.
...
+     utility function. Be sure to include the `file_hash` argument, which allows
+     cache invalidation on the downloaded weights. The command line programs
+     `shasum` and `sha256sum` can compute a file hash.
```
Comment (fchollet) [inline CONTRIBUTING.md:236]: You verified that the output is the same? I wasn't sure

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -211,3 +211,39 @@ command above.
...
+ - You should help us verify that the accuracy of the model with pre-trained
+   weighted matches the reported results of the cited paper.
+ - You should add any new applications to the unit tests defined in
+   `applications_test.py` and `applications_load_weight_test.py`.
+ - Avoid putting computation in the `preprocess_input` function. For new
```
Comment (fchollet) [inline CONTRIBUTING.md:241]: Avoid putting -> Don't put (this is imperative)

Add parens: `preprocess_input()`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -211,3 +211,39 @@ command above.
...
+   applications `preprocess_input` should pass through user inputs unaltered.
```
Comment (fchollet) [inline CONTRIBUTING.md:242]: Add parens

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -211,3 +211,39 @@ command above.
...
+   Prefer writing application that can handle raw inputs directly.
+   [Preprocessing layers](https://keras.io/guides/preprocessing_layers/) may
+   help with this.
```
Comment (fchollet) [inline CONTRIBUTING.md:245]: Mention `Rescaling(1. / 255.)` since typically this is the only layer needed

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -211,3 +211,39 @@ command above.
...
+ - You shoud include a companion pull request to the keras.io
```
Comment (fchollet) [inline CONTRIBUTING.md:246]: "Once the PR is approved, you should create" (no need to update keras.io until we're actually publishing the new model)

**Review by fchollet — COMMENTED** [review]
Awesome, thanks for the PR!

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -211,3 +211,42 @@ command above.
...
+## Contributing to Keras applications
+
+Contributions to the
+[pre-trained application library](https://keras.io/api/applications/)
+are welcome. Code for Keras applications is located in Keras repository in
+[keras/applications](https://github.com/keras-team/keras/blob/master/keras/applications).
+When contributing to Keras applications, please keep following checklist in
+mind.
+
+ - Keras applications must implement an established and widely used model.
```
Comment (innat) [inline CONTRIBUTING.md:224]: @mattdangerw 
How is the term **...widely used model** defined? Is it based on only the number of citations of the model used in well-reputed conferences/journals? In that case, I think the conf/journal publisher should be realized also. 20 citations from random journals/publishers would be easy to find these days.

Comment (mattdangerw) [reply CONTRIBUTING.md:224]: Yeah, @fchollet could probably add more color. I don't think there is any fool proof way of defining widely used. 20 citations is just a minimum bar really. We can't enumerate all reputable journals.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -211,3 +211,42 @@ command above.
...
+     merging the pull request.
+   - Weights should be downloaded with the
+     [get_file()](https://keras.io/api/utils/python_utils/#getfile-function)
+     utility function. Be sure to include the `file_hash` argument, which allows
+     cache invalidation on the downloaded weights. The command line programs
+     `shasum` and `sha256sum` can compute a file hash.
+ - You should help us verify that the accuracy of the model with pre-trained
+   weighted matches the reported results of the cited paper.
+ - You should add any new applications to the unit tests defined in
+   `applications_test.py` and `applications_load_weight_test.py`.
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply CONTRIBUTING.md:240]: No other applications do. We should probably have an example if we are going to ask for this?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -211,3 +211,42 @@ command above.
...
+ - For backwards compatibility, all applications should provide a
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply CONTRIBUTING.md:241]: Our load weights tests do. Not sure about anything else.

File: CONTRIBUTING.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (rchao) [reply CONTRIBUTING.md:240]: I see. Not necessarily asking for one here, but I was curious that if there's no unit test in place I wonder how people make sure their code runs.

[the remark this answers is not in the document]
Comment (mattdangerw) [reply CONTRIBUTING.md:240]: I believe the load weights test will load the weights, call predict, and make sure things run.

But I'm also definitely not the expert on Keras applications, so may be missing things. This is just me typing up the notes I had from doing at application review with @fchollet.

File: CONTRIBUTING.md (hunk 8 above)
Comment (fchollet) [reply CONTRIBUTING.md:224]: We don't want to set too high a bar -- 20 citations is enough to filter out the things we want to exclude. It's really more about excluding the clutter than about formally defining "widely used".

File: CONTRIBUTING.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply CONTRIBUTING.md:240]: That's right, but the test is really more a sanity check that verifies nothing is obviously wrong -- it does not check that the validation accuracy of the model matches what's expected. It would pass just as well for a 60% accuracy model and for a 75% accuracy model. This relies on the fact that the outcome of loading weights in a model is binary: either it works as expected, or nothing makes sense. It's extremely rare for the process to "most work but not quite" (the only possibility is that the data normalization step is off by a bit -- any other bug will lead to catastrophic failure).

In practice we need a one-time check (on our side) that the validation accuracy matches what's expected, after the model is submitted. We can always revert in case of major issue.

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
